Avoid counting lines in source maps

nsavoire opened the pull request:
Added a `countLines` parameter to `scanner.scan`, so that scanner does not count number of lines in source map since this information is not used and the process could lead to unexpected CPU consumption.

Qard reviewed (changes requested):
I would skip making it configurable. _Should_ just be a matter of deleting the `byLine` assignment to `reader!.pipe(split())` and changing all references to the `byLine` stream to just use the `reader` stream instead, and then remove any line-related stuff from the resolved object and the counting logic.

nsavoire commented:
> I would skip making it configurable. _Should_ just be a matter of deleting the `byLine` assignment to `reader!.pipe(split())` and changing all references to the `byLine` stream to just use the `reader` stream instead, and then remove any line-related stuff from the resolved object and the counting logic.

Removing all line-related stuff implies also removing optional hashing (which is not used). At this point, all scanner.ts is doing is walking a directory tree, and I wonder if there is a better / more standard alternative to do that in node.js ?

nsavoire commented:
> > I would skip making it configurable. _Should_ just be a matter of deleting the `byLine` assignment to `reader!.pipe(split())` and changing all references to the `byLine` stream to just use the `reader` stream instead, and then remove any line-related stuff from the resolved object and the counting logic.
> 
> If we remove all line-related stuff, then we could also remove the optional hashing (which is not used), and avoid reading files entirely. At this point, all scanner.ts is doing is walking a directory tree, and I wonder if there is a better / more standard alternative to do that in node.js ?

I removed scanner.ts altogether and replaced it with readdirp.

Qard commented on ts/src/sourcemapper/sourcemapper.ts:
```diff
@@ -250,7 +249,13 @@ async function createFromMapFiles(mapFiles: string[]): Promise<SourceMapper> {
 async function getMapFiles(baseDir: string): Promise<string[]> {
-  const fileStats = await scanner.scan(false, baseDir, /.js.map$/);
-  const mapFiles = fileStats.selectFiles(/.js.map$/, process.cwd());
+  const mapFiles: string[] = [];
+  for await (const entry of readdirp(baseDir, {
+    fileFilter: '*.js.map',
+    directoryFilter: ['!.git', '!node_modules'],
+    alwaysStat: true,
+  })) {
+    mapFiles.push(entry.path);
+  }
```
This is not `quite` the same as `entry.path` is relative to `baseDir` rather than `process.cwd()`. This happens to still work just because dd-trace-js is passing `[ process.cwd() ]` as the directory list for `SourceMapper.create`, but it would be a breaking change for anyone using any _other_ input to that. It also renders the loop logic in `SourceMapper.create` irrelevant. We should probably either mark this as a semver major change and also update `SourceMapper.create` to just take a single directory string, or update this to have the same relative pathing behaviour as before.

nsavoire replied:
> This is not `quite` the same as `entry.path` is relative to `baseDir` rather than `process.cwd()`. This happens to still work just because dd-trace-js is passing `[ process.cwd() ]` as the directory list for `SourceMapper.create`, but it would be a breaking change for anyone using any _other_ input to that. It also renders the loop logic in `SourceMapper.create` irrelevant. We should probably either mark this as a semver major change and also update `SourceMapper.create` to just take a single directory string, or update this to have the same relative pathing behaviour as before.

Good catch! But the initial behaviour looks very suspicious to me:
* `getMapFiles` returns paths relative to `process.cwd()` (which is not even always true, because I have the impression that `selectFiles` produces relative path only if `process.cwd()` is a prefix of map files paths).
* `SourceMapper.create` then recomputes absolute paths with `path.resolve(dir, mapFile)` ie. by using `baseDir` as prefix. This looks wrong to me as this would produce incorrect paths when `baseDir` is different from `process.cwd()`.

Qard replied:
Possibly? I didn't look super closely at it, but perhaps worth adding some tests to make sure the way things work is reasonable, and if it's producing different results then this should be marked as a major change.

It's _fine_ if this is marked as major, we can still upgrade it without being a major in dd-trace as we won't ever encounter baseDir being different, but we should play nice with any other users that might depend on this module as there _are_ a few. https://www.npmjs.com/package/@datadog/pprof?activeTab=dependents

nsavoire replied:
Note that `getMapFiles` is not exported, only `SourceMapper` is.
Here are the results of my tests
* if baseDir is `cwd`, then returned results of `getMapFiles` are the same
* if baseDir is different from `cwd`, then:
  * if mapFile is not in `cwd` sub-tree, `getMapFiles` would previously return absolute paths and now will return relative paths, but since paths are later on resolved with `baseDir`, the end results will be the same
  * if map file is in `cwd` sub-tree, `getMapFiles` would previously return relative paths to `cwd` that would then be resolved with `baseDir` leading to invalid paths. Now it will return relative paths to `baseDir` and  resolve to valid path with `baseDir`. The result is different but is more of a bug fix than an API change.

Qard replied:
Okay, I think it could be reasonable to change as a bugfix then. 🤔

nsavoire replied:
I removed `readdirp` and implemented a recursive directory walk (without following symlinks, similar to what was done with scanner.ts/findit2).

Qard commented on ts/src/sourcemapper/sourcemapper.ts:
```diff
@@ -249,8 +247,57 @@ async function createFromMapFiles(mapFiles: string[]): Promise<SourceMapper> {
 }
 
+function isErrnoException(e: unknown): e is NodeJS.ErrnoException {
+  if (e instanceof Error && 'code' in e) return true;
+  else return false;
```
This could be simplified to:

```suggestion
  return e instanceof Error && 'code' in e;
```

Qard reviewed (approved):
One minor nit, other than that LGTM. 👍🏻
